### DISCUSSION TITLE
Documentation for services

### DISCUSSION THREAD

**Discussion by girishnuli** [directus/directus#16780] [Archive]
I am exploring [hooks](https://docs.directus.io/extensions/hooks.html), but having a hard time finding the documentation for the injected services like schema, database, accountability etc. Is there any API reference available that describes how to use these services?

**Comment by paescuj** [comment]
I'm sure the docs on this matter will be expanded sooner or later (see issues / pull requests in [directus/docs](https://github.com/directus/docs)). In the meantime you can use other resources, like examples from [discussions](https://github.com/directus/directus/discussions), help in [Discord](https://directus.chat), etc.

**Reply by girishnuli** [reply]
Thanks, I will checkout the discord. Hopefully the docs will come up soon as it is a major limitation to writing custom APIs right now. We are keen to use directus as a full fledged backend, but can use it only as a CMS till the services docs are up.

**Reply by paescuj** [reply]
One thing I'd like to mention is that there is the possibility to [sponsor specific features](https://docs.directus.io/getting-started/support.html#sponsored-work) which will then be prioritized.

**Comment by raphourbe** [comment]
Hi,

Here is an example in an Endpoint, but it should work in a Hook too:

```js:index.js
export default (router, { services, exceptions, database, schema, logger }) => {
  router.post('/', async (req, res) => {
    const { ItemsService, MailService } = services;
    const { ServiceUnavailableException } = exceptions;

    const customerDetailsService = new ItemsService("customer_details", {
      schema: req.schema,
    });
    const customerSubscriptionService = new ItemsService(
      "customer_subscriptions",
      {
        schema: req.schema,
        accountability: {
          role: "admin",
          admin: true,
        },
      }
    );
   customerSubscriptionService.createOne({
          customer_details: "something",
          subscription_items: "something",
          trial_ends_at:  "something",
        })
    const mailer = new MailService({ schema, knex: database });
    // ...
  }
}
```

You can check the **services** folder of the [repo](https://github.com/directus/directus/tree/main/api/src/services) to see what is available in the `services`, and what functions are available to each service.

You can use `database` and `schema` to instantiate the Service. I found that using the `ItemsService` is what you really need. In my example, I use it to call the `customer_details` table in my database and my `customer_subscriptions` table. For the last one, you can see how to use accountability inside.

The MailService allows you to send mail using NodeMailer under the hood:

```js
await mailer .send({
  to: customerDetails.emaill,
  bcc: 'mymail@mail.com',
  subject: 'Hi !',
  template: {
    name: 'subscription-started',
    data: {
      firstName: customerDetails.user_id.first_name,
      trialEnd: trialEnd,
      daysOfTrial: subscriptionItem.days_of_trial,
    },
  },
})
```

You can also have `accountability` given directly and make check with it (in hooks, accountability is given differently, but should work the same):

```js
router.post("/", async (req, res, next) => {
  // admin or customer
  if (
    req.accountability?.admin ||
    req.accountability?.role === process.env.ROLE_ID
  ) {
   // do something 
}
 else { return; }
}
```

I have done a full how-to to create a Saas with Directus and Stripe. In the [last post](https://rphl.dev/blog/building-saas-with-directus-and-stripe-part4#stripe-configuration), I show how to implement an Endpoint to handle Stripe webhooks. Some of the examples above come from it. There is not a part for a detailed Hook extension though, but you will see how to use `services`.

Hope this help.

**Reply by girishnuli** [reply]
Thank you for the detailed reply! Really appreciate it. I feel that the repo and your blog should give us enough info to understand services better.
